viewer/Viewer.cpp

```diff
@@ -106,26 +107,55 @@ namespace Enki
 		yaw(yaw),
 		pitch(pitch)
 	{
-		
+		update(false);
 	}
 	
+	void ViewerWidget::CameraPose::update(bool trackballMode, QVector3D targetPosition, float zNear)
+	{
+		forward = QVector3D( cos(yaw)*cos(pitch), sin(yaw)*cos(pitch), sin(pitch) );
+		left = QVector3D::crossProduct(QVector3D(0,0,1), forward).normalized();
+		up = QVector3D::crossProduct(forward, left).normalized();
+
+		if(trackballMode)
+		{
+			pos.rx() = targetPosition.x() - radius*forward.x();
+			pos.ry() = targetPosition.y() - radius*forward.y();
+			altitude = targetPosition.z() + zNear*1.01 - radius*forward.z();
+		}
+	}
+
 	ViewerWidget::ViewerWidget(World *world, QWidget *parent) :
 		QGLWidget(parent),
 		timerPeriodMs(30),
-		world(world),
-		worldList(0),
-		mouseGrabbed(false),
 		camera(CameraPose(
 			QPointF(world->w * 0.5, qMax(0., world->r)),
 			qMax(qMax(world->w, world->h), world->r*2) * 0.85,
 			M_PI/2,
 			-(3*M_PI)/8
 		)),
-		wallsHeight(10),
 		doDumpFrames(false),
+		world(world),
+		worldList(0),
+		mouseGrabbed(false),
+		wallsHeight(10),
+		trackballView(false),
 		dumpFramesCounter(0)
 	{
 		initTexturesResources();
+		pointedObject = 0;
+		selectedObject = 0;
+
+		controlError1 = "object translation not avalaible in trackball mode";
```

Comment:
Ok, I see your need. So you want to be able to display messages that stay some times on the screen, with some that do not allow duplication. You use the string itself to detect this duplication. I think that this rationale makes sense!

In the same time non-translated strings here are really useless in applications. So let's think how we can improve the design:

1. Let's pass the strings through a `tr()` function, to get them translated (see http://doc.qt.io/qt-4.8/i18n-source-translation.html)
2. Let's not keep `controlError1` and `controlErro2` as members of the class, as anyway you emit them in a single place, but rather check the duplication in `sendMessage()`.
3. Rename `sendMessage()` to `addErrorMessage()` and put it in a protected function rather than a public slot.